File: backend/GraficaModerna.Domain/Entities/Coupon.cs
namespace GraficaModerna.Domain.Entities;

public class Coupon
{

    public Coupon()
    {
    }

    public Coupon(string code, decimal percentage, int daysValid)
    {
        Id = Guid.NewGuid();
        Code = code.ToUpper().Trim();
        DiscountPercentage = percentage;
        ExpiryDate = DateTime.UtcNow.AddDays(daysValid);
        IsActive = true;
    }

    public Guid Id { get; set; }
    public string Code { get; set; } = string.Empty;
    public decimal DiscountPercentage { get; set; } 
    public DateTime ExpiryDate { get; set; }
    public bool IsActive { get; set; } = true;

    public bool IsValid()
    {
        return IsActive && DateTime.UtcNow <= ExpiryDate;
    }
}

File: app/Models/Coupon.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Concerns\HasUuids;

class Coupon extends Model
{
    use HasUuids;

    protected $fillable = ['code', 'discount_percentage', 'expiry_date', 'is_active'];

    protected $casts = [
        'expiry_date' => 'datetime',
        'is_active' => 'boolean',
        'discount_percentage' => 'decimal:2'
    ];

    public function isValid(): bool
    {
        return $this->is_active && $this->expiry_date > now();
    }
}